x/concentrated-liquidity/position.go

```diff
@@ -249,6 +251,81 @@ func (k Keeper) CreateFullRangePosition(ctx sdk.Context, concentratedPool types.
+	// Create a full range (min to max tick) concentrated liquidity position.
+	positionId, amount0, amount1, liquidity, joinTime, err = k.CreateFullRangePosition(ctx, concentratedPool, owner, coins)
+	if err != nil {
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, err
+	}
+
+	// Mint cl shares for the position and lock them for the remaining lock duration.
+	// Also sets the position ID to underlying lock ID mapping.
+	concentratedLockId, _, err := k.MintSharesLockAndUpdate(ctx, concentratedPool, positionId, owner, remainingLockDuration, liquidity)
+	if err != nil {
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, err
+	}
+
+	return positionId, amount0, amount1, liquidity, joinTime, concentratedLockId, nil
+}
+
+// CreateFullRangePositionUnlocking creates a full range (min to max tick) concentrated liquidity position for the given pool ID, owner, and coins.
+// This function is strictly used when migrating a balancer position to CL, where the balancer position is currently unlocking.
+// We lock the cl position for whatever the remaining time is from the balancer position and immediately begin unlocking from where it left off.
+func (k Keeper) CreateFullRangePositionUnlocking(ctx sdk.Context, concentratedPool types.ConcentratedPoolExtension, owner sdk.AccAddress, coins sdk.Coins, remainingLockDuration time.Duration) (positionId uint64, amount0, amount1 sdk.Int, liquidity sdk.Dec, joinTime time.Time, concentratedLockID uint64, err error) {
+	// Create a full range (min to max tick) concentrated liquidity position.
+	positionId, amount0, amount1, liquidity, joinTime, err = k.CreateFullRangePosition(ctx, concentratedPool, owner, coins)
+	if err != nil {
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, err
+	}
+
+	// Mint cl shares for the position and lock them for the remaining lock duration.
+	// Also sets the position ID to underlying lock ID mapping.
+	concentratedLockId, underlyingLiquidityTokenized, err := k.MintSharesLockAndUpdate(ctx, concentratedPool, positionId, owner, remainingLockDuration, liquidity)
+	if err != nil {
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, err
+	}
+
+	// Begin unlocking the newly created concentrated lock.
+	concentratedLockID, err = k.lockupKeeper.BeginForceUnlock(ctx, concentratedLockId, underlyingLiquidityTokenized)
+	if err != nil {
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, err
+	}
+
+	return positionId, amount0, amount1, liquidity, joinTime, concentratedLockID, nil
+}
+
+// MintSharesLockAndUpdate mints the shares for the concentrated liquidity position and locks them for the given duration. It also updates the position ID to underlying lock ID mapping.
```

Comment:
Can you elaborate on why minting needs to happen?

I think this is only for superfluid purposes.

Sidenote: could you please kick off a docs PR so that we can port over such details there as we continue to review

Reply:
https://github.com/osmosis-labs/osmosis/pull/4883/commits/eb21750d9d508ba18854f505f5665b9413849b47

Reply:
Could you please expand on the reason for minting that this is so that the bank invariants don't break as we send these coins to lockup module account

Also, bump on starting up a doc on these details

Reply:
https://github.com/osmosis-labs/osmosis/pull/4883/commits/894b61990cd6c5e60c0651acf4c22df0a2988ccc